scripts/check-title-labels.lean

```diff
@@ -0,0 +1,267 @@
+-- FIXME: this should be added to batteries, as a version of "splitOn"
+def splitAtPos (s : String) (pos : String.Pos) : String × String :=
+  let before := s.extract 0 pos
+  let after := s.extract (s.next pos) s.endPos
+  (before, after)
+
+/--
+Check if `title` matches the mathlib conventions for PR titles
+(documented at https://leanprover-community.github.io/contribute/commit.html).
+Not all checks are implemented: for instance, no effort is made to verify if the title or body
+are written in present imperative tense.
+
+Return all error messages for violations found.
+-/
+def validateTitle (title: String) : Array String := Id.run do
+  -- The title should be of the form "abbrev: main title" or "abbrev(scope): main title".
+  if !title.contains ':' then
+    return #["error: the PR title does not contain a colon"]
+
+  -- Split 'title' at the first occurrence of ':'.
+  let (type, main) := splitAtPos title (title.find (· == ':'))
+  let mut errors := #[]
+  if main.endsWith "." then
+    errors := errors.push "error: the PR title should not end with a full stop"
+
+  if main.front != ' ' then
+    errors := errors.push "error: the PR should have the form 'abbrev: main title', with a space"
+  else if main.containsSubstr "  " then
+    errors := errors.push "error: the PR title contains multiple consecutive spaces; please add just one"
+  let main := main.removeLeadingSpaces
+  if main.front.toLower != main.front then
+    errors := errors.push "error: the main PR title should be lowercased"
+
+  -- `type` should be of the form abbrev or abbrev(scope), where `scope` is of the form
+  -- `Algebra/Basic`, without a leading `Mathlib` or a trailing `.lean`
+  let N := (type.splitOn "(").length
+  let M := (type.splitOn ")").length
+  let known_types := ["feat", "chore", "perf", "refactor", "style", "fix", "doc", "test"]
+  if (N, M) == (1, 1) then
+    if !known_types.contains type then
+      errors := errors.push s!"error: the PR type should be one of {", ".intercalate (known_types.map (s!"\"{·}\""))}"
+  else if (N, M) == (2, 2) then
+    if !type.endsWith ")" then
+      errors := errors.push s!"error: the PR type should be of the form abbrev(scope)"
+    let idx := type.find (· == '(')
+    let idx2 := type.find (· == ')')
+    if idx2 < idx then
+      errors := errors.push "error: mismatched parentheses; the PR type should be of the form abbrev(scope)"
+    let (type, scope) := (splitAtPos type idx)
+    let scope := scope.dropRight 1
+    if !known_types.contains type then
+      errors := errors.push s!"error: the PR type should be one of {", ".intercalate (known_types.map (s!"\"{·}\""))}"
+    if scope.startsWith "Mathlib/" then
+      errors := errors.push s!"error: the PR scope must not start with 'Mathlib/'"
+    if scope.endsWith ".lean" then
+      errors := errors.push s!"error: a PR's scope must not end with '.lean'"
+    -- Future: we could further validate the scope, that this is a valid module or directory
```

Comment:
Perhaps also check whether scope is `/` separated or `.` separated?

Reply:
Good idea, done.